bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -0,0 +1,531 @@
+        return decryptedToken;
+    }
+
+    private String encrypt(String token) throws GeneralSecurityException {
+        if (!storageCipher.isPresent()) {
+            return token; // do nothing if no cipher
+        }
+        return storageCipher.get().encrypt(token);
+    }
+
+    private void loadAllHandlesAndTokens() {
+        // load all handles from store
+        allHandles = storageFacade.getAllHandlesFromIndex();
+
+        for (String handle : allHandles) {
+            allHandles.add(handle);
+            logger.debug("Found previously persisted handle: {}", handle);
+            String storageKeyAccessToken = ACCESS_TOKEN_RESPONSE.getKey(handle);
+
+            AccessTokenResponse accessTokenResponseFromStore = (AccessTokenResponse) storageFacade.get(handle,
+                    ACCESS_TOKEN_RESPONSE);
+            try {
+                if (accessTokenResponseFromStore == null) {
+                    // token does not exist
+                    continue;
+                }
+
+                AccessTokenResponse decryptedAccessToken = decryptToken(accessTokenResponseFromStore);
+
+                accessTokenCache.put(storageKeyAccessToken, decryptedAccessToken);
+            } catch (GeneralSecurityException e) {
+                logger.error("Unable to decrypt token, using it directly.", e);
+                accessTokenCache.put(storageKeyAccessToken, accessTokenResponseFromStore);
+            } catch (IllegalArgumentException e) {
+                logger.error("Unable to decrypt token, using it directly.", e);
+                accessTokenCache.put(storageKeyAccessToken, accessTokenResponseFromStore);
+            }
+        }
+    }
+
+    /**
+     * Reference is optional -- basically the storage service may be removed during runtime.
+     * When it is removed, the service should still stay up for read-only operations.
+     *
+     * Using default ReferencePolicyOption.RELUCTANT, so it should not be called twice.
+     *
+     * @param storageService
+     */
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC)
+    protected synchronized void setStorageService(StorageService storageService) {
+        if (_storageServiceDoNotUseDirectly == null) {
+            _storageServiceDoNotUseDirectly = storageService;
+            _storeDoNotUseDirectly = storageService.getStorage(STORE_NAME);
+            storageFacade = new StorageFacade(this._storeDoNotUseDirectly);
+        }
+    }
+
+    protected synchronized void unsetStorageService(StorageService storageService) {
+        if (_storageServiceDoNotUseDirectly == storageService) {
+            _storageServiceDoNotUseDirectly = null;
+            storageFacade.close();
+        }
+    }
+
+    /**
+     * Static policy -- dont want to change cipher on the fly!
+     * There may be multiple storage ciphers, choose the one that matches the target (done at activate)
+     *
+     * @param storageCipher
```

Comment:
empty param description can be removed

Reply:
it's automatically generated by using /**